odxtools/parameters/tablekeyparameter.py

```diff
@@ -37,37 +40,107 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+
+        self._table_row: Optional[TableRow] = None
         if self.table_row_ref:
-            self._table_row = odxlinks.resolve(self.table_row_ref)
+            if TYPE_CHECKING:
+                self._table_row = odxlinks.resolve(self.table_row_ref, TableRow)
+            else:
+                self._table_row = odxlinks.resolve(self.table_row_ref)
 
     def _resolve_snrefs(self, diag_layer: "DiagLayer") -> None:
         super()._resolve_snrefs(diag_layer)
 
         if self.table_snref:
-            self._table = diag_layer.tables[self.table_snref]
+            ddd_spec = diag_layer.diag_data_dictionary_spec
+            self._table = ddd_spec.tables[self.table_snref]
         if self.table_row_snref:
-            self._table_row = diag_layer.tables[self.table_row_snref]
+            # make sure that we know the table to which the table row
+            # SNREF is relative to.
+            assert self._table is not None, "If a table-row short name " \
+                "reference is defined, a table must also be specified."
+            self._table_row = self._table.table_rows[self.table_row_snref]
 
     @property
     def table(self) -> "Table":
         return self._table
 
     @property
-    def table_row(self) -> "TableRow":
+    def table_row(self) -> Optional["TableRow"]:
         return self._table_row
 
     def is_required(self):
-        raise NotImplementedError("TableKeyParameter.is_required is not implemented yet.")
+        return self._table_row is None
 
     def is_optional(self):
-        raise NotImplementedError("TableKeyParameter.is_optional is not implemented yet.")
-
-    def get_coded_value(self):
-        raise NotImplementedError("Encoding a TableKeyParameter is not implemented yet.")
-
-    def get_coded_value_as_bytes(self):
-        raise NotImplementedError("Encoding a TableKeyParameter is not implemented yet.")
-
-    def decode_from_pdu(self, coded_message, default_byte_position=None):
-        raise NotImplementedError("Decoding a TableKeyParameter is not implemented yet.")
+        return not is_required
+
+    def get_coded_value(self, physical_value=None):
+        key_dop = self.table.key_dop
+        return key_dop.convert_physical_to_internal(physical_value)
+
+    def get_coded_value_as_bytes(self, encode_state: EncodeState) -> bytes:
+        tr = encode_state.table_keys[self.short_name]
+        key_dop = self.table.key_dop
+        bit_position = 0 if self.bit_position is None else self.bit_position
+        if key_dop is None:
+            raise EncodeError(f"Table '{self.table.short_name}' does not define "
+                              f"a KEY-DOP, but is used in TABLE-KEY parameter "
+                              f"'{self.short_name}'")
+        return key_dop.convert_physical_to_bytes(tr.key, encode_state, bit_position=bit_position)
+
+    def encode_into_pdu(self, encode_state: EncodeState) -> bytes:
+        # Set the value of the length key in the length key dict.
```

Comment:
length key?

Reply:
copy and paste: Best friend, worst enemy...

fixed, thanks